### DISCUSSION TITLE
Artifacts in the inscriptions on the contour lines.

### DISCUSSION THREAD

**Discussion by netman220** [mapsforge/mapsforge#1297] [Q&A] [answered]
Hi.
I use the Mapsforge library (Version 0.16.0) with Cruiser (Version 2.1.2). The map is of the Bulgarian mountains ( https://map.bgmountains.org/ ) with a custom theme and map is version 3. The map is displayed with artifacts that are visible in attached png files. Artifacts are always on the borders of the tiles. In Oruxmaps 7.4.25, which has an old version of the Mapsforge library, the artifacts are missing. My question is what could be the cause of these artifacts?
![20220224_155540](https://user-images.githubusercontent.com/100360351/155540352-4d73fef4-fe76-4743-9259-cf936b85413f.png)
![20220224_155805](https://user-images.githubusercontent.com/100360351/155540362-d14c7808-01fd-44be-abb9-a8573b270455.png)
![20220224_105001](https://user-images.githubusercontent.com/100360351/155540363-717e436d-367e-4269-a6ad-7699718ad59b.png)

**Comment by devemux86** [comment]
Are they elevation labels from some contour line?
For anyone to check them, they need a map file and map style.

You can test the contours of [OpenAndroMaps](https://www.openandromaps.org/en) map files with [Elevate](https://www.openandromaps.org/en/legend/elevate-mountain-hike-theme) map style
and see how they work and are written.

**Reply by netman220** [reply]
Hi.
[Download Theme](http://samsiev.eu/Users/Maps/BGM_theme.zip)
[Download Map](http://samsiev.eu/Users/Maps/bgmountains.map)
With OpenAndroMaps map and Elevate map style everything is fine, but there is a completely different method used to show the contour lines. There the inscriptions on the contour lines are every 100 meters and the contour lines are every 20 meters. 
Location of the artifacts (the above screenshots):
1. 42.7311 N, 25.05008 E; zoom level 16
2. 42.73074 N, 24.96983; zoom level 17
3. 43.12399 N,  25.66516 E; zoom level 16

**Comment by devemux86** [comment]
> With OpenAndroMaps map and Elevate map style everything is fine

Then better compare the map creation and theme rules with OpenAndroMaps
and see how they need to be done to work properly.

**Reply by netman220** [reply]
Hello
The map and theme of BGmountains are completely different from those of OSM. The theme uses only the Elevate theme format. The content is radically different. Please review the part of the file that is responsible for the contour lines. I will be glad if you find the reason for these artifacts. 

=================== CODE from Theme =====================

<rule e="any" k="*" v="*" cat="contour">

    <rule e="way" k="c1000" v="yes" zoom-min="10">
        <line stroke="#40330000" stroke-width="0.6" stroke-linecap="butt" />

        <rule e="way" k="*" v="*" zoom-min="10" zoom-max="13">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="50" repeat-gap="400"  font-size="8" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="14" zoom-max="15">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="50" repeat-gap="500"  font-size="8" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="16" zoom-max="17">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="100" repeat-gap="600" font-size="13" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="18" zoom-max="18">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="150" repeat-gap="700"  font-size="15" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="19">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="200" repeat-gap="800"  font-size="17" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
    </rule>

	<rule e="way" k="c500" v="yes" zoom-min="11">
		<line stroke="#40330000" stroke-width="0.6" stroke-linecap="butt" />
        
        <rule e="way" k="*" v="*" zoom-min="11" zoom-max="13">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="50" repeat-gap="400"  font-size="8" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="14" zoom-max="15">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="50" repeat-gap="500"  font-size="8" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="16" zoom-max="17">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="100" repeat-gap="600"  font-size="13" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="18" zoom-max="18">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="150" repeat-gap="700"  font-size="15" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="19">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="200" repeat-gap="800"  font-size="17" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
	</rule>

	<rule e="way" k="c200" v="yes" zoom-min="12" >
		<line stroke="#40330000" stroke-width="0.6" stroke-linecap="butt" />

        <rule e="way" k="*" v="*" zoom-min="12" zoom-max="13">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="50" repeat-gap="400"  font-size="8" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="14" zoom-max="15">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="50" repeat-gap="500"  font-size="8" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="16" zoom-max="17">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="100" repeat-gap="600"  font-size="13" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="18" zoom-max="18">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="150" repeat-gap="700"  font-size="15" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="19">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="200" repeat-gap="800"  font-size="17" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
	</rule>

    <rule e="way" k="c100" v="yes" zoom-min="14">
        <line stroke="#40330000" stroke-width="0.6" stroke-linecap="butt" />
        <rule e="way" k="*" v="*" zoom-min="14" zoom-max="15">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="50" repeat-gap="500"  font-size="8" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="16" zoom-max="17">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="100" repeat-gap="600"  font-size="13" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="18" zoom-max="18">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="150" repeat-gap="700"  font-size="15" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
        <rule e="way" k="*" v="*" zoom-min="19">
            <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="200" repeat-gap="800"  font-size="17" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
        </rule>
    </rule>

	<rule e="way" k="c50" v="yes" zoom-min="15" zoom-max="15">
		<line stroke="#40330000" stroke-width="0.3" stroke-linecap="butt" />
        <pathText k="ref" display="always" priority="10" font-style="bold" repeat-start="50" repeat-gap="500" font-size="8" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" /> 
	</rule>


	<rule e="way" k="c20" v="yes" zoom-min="16">
		<line stroke="#40330000" stroke-width="0.3" stroke-linecap="butt" />
		<rule e="way" k="*" v="*" zoom-min="16" zoom-max="16">
			<pathText k="ref" display="always" font-style="bold" repeat-start="50" repeat-gap="500" font-size="9" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
		</rule>
		<rule e="way" k="*" v="*" zoom-min="17" zoom-max="17">
			<pathText k="ref" display="always" font-style="bold" repeat-start="100" repeat-gap="600" font-size="13" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
		</rule>
		<rule e="way" k="*" v="*" zoom-min="18" zoom-max="18">
			<pathText k="ref" display="always" font-style="bold" repeat-start="150" repeat-gap="700" font-size="15" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
		</rule>
		<rule e="way" k="*" v="*" zoom-min="19">
			<pathText k="ref" display="always" font-style="bold" repeat-start="200" repeat-gap="800" font-size="17" fill="#A0330000" stroke="#A0FFFFFF" stroke-width="8" />
		</rule> 
	</rule>

</rule>

**Comment by devemux86** [comment]
Please study the [render theme documentation](https://github.com/mapsforge/mapsforge/blob/master/docs/Rendertheme.md) and the OpenAndroMaps [tag-mapping](https://www.openandromaps.org/en/map-basics-2/tag-mapping) and [Elevate](https://www.openandromaps.org/en/legend/elevate-mountain-hike-theme) theme.

_**Use proper key + value pairs in map creation and theme: use one common key with many values**._

**Comment by netman220** [comment]
Hello
For a test, I simplified the section of contour lines. I only show 20 m contour lines. The artifacts remain. If I disable the line with "pathText" the elevation is not displayed and the artifacts disappear, which means that they come from this line. I noticed that the distance from the artifact to the border of the tile is 20 meters.
I enclose the code and a screenshot (42.73142 N, 25.04917 E; zoom level 18).
### [Link to Code](http://samsiev.eu/Users/Maps/ContourLinesTest.txt)

![20220225_174928](https://user-images.githubusercontent.com/100360351/155758245-e05af190-eb41-4fdc-b740-56aa8db7cc90.png)

**Comment by devemux86** [comment, marked as answer]
As mentioned [above](https://github.com/mapsforge/mapsforge/discussions/1297#discussioncomment-2250111):

> Please study the [render theme documentation](https://github.com/mapsforge/mapsforge/blob/master/docs/Rendertheme.md) and the OpenAndroMaps [tag-mapping](https://www.openandromaps.org/en/map-basics-2/tag-mapping) and [Elevate](https://www.openandromaps.org/en/legend/elevate-mountain-hike-theme) theme.

> _**Use proper key + value pairs in map creation and theme: use one common key with many values**._

---

OpenAndroMaps uses the [tag-mapping](https://www.openandromaps.org/wp-content/snippets/makes/tagmapping-urban.xml) in map creation:
```xml
<!-- Contourlines -->
<osm-tag key="contour_ext" value="elevation_major" force-polygon-line="true" zoom-appear="9"/>
<osm-tag key="contour_ext" value="elevation_medium" force-polygon-line="true" zoom-appear="10"/>
<osm-tag key="contour_ext" value="elevation_minor" force-polygon-line="true" zoom-appear="12"/>
```
And the [Elevate](https://www.openandromaps.org/en/legend/elevate-mountain-hike-theme) theme uses the rules:
```xml
<rule e="way" k="contour_ext" v="elevation_major">
<rule e="way" k="contour_ext" v="elevation_medium">
<rule e="way" k="contour_ext" v="elevation_minor">
```

---

You use c1000=yes, c500=yes, ...

Change the tag-mapping in map creation:
```xml
<osm-tag key="contour" value="c1000" force-polygon-line="true" zoom-appear="10"/>
<osm-tag key="contour" value="c500" force-polygon-line="true" zoom-appear="11"/>
<osm-tag key="contour" value="c200" force-polygon-line="true" zoom-appear="12"/>
<osm-tag key="contour" value="c100" force-polygon-line="true" zoom-appear="14"/>
<osm-tag key="contour" value="c50" force-polygon-line="true" zoom-appear="15"/>
<osm-tag key="contour" value="c20" force-polygon-line="true" zoom-appear="16"/>
```
Then change the render theme:
```xml
<rule e="way" k="contour" v="c1000">
...
```

**Comment by DeyanVasilev** [comment]
Hello Emux,

it's Deyan from BGMountains. Thanks a lot for helping us resolve this issue.
All contours are tagged with key/value pairs being "c1000=yes", "c500=yes", etc and not with k/v "contour=c1000"

My tag-mapping.xml has been modified as follows:

>
    
    <ways>
        
        <osm-tag key="c20" value="yes" zoom-appear="10" />
        <osm-tag key="c50" value="yes" zoom-appear="10" />
        <osm-tag key="c100" value="yes" zoom-appear="10" />
        <osm-tag key="c200" value="yes" zoom-appear="10" />
        <osm-tag key="c500" value="yes" zoom-appear="10" />
        <osm-tag key="c1000" value="yes" zoom-appear="10" />
        <osm-tag key="contour_ele" value="%f" zoom-appear="10" />
    </ways>

Our problem as mentioned by netman220 is present only on apps that use newer mapsforge library. On older apps with older mapsforge this problem is non-existent.

**Reply by devemux86** [reply]
You probably also need the [force-polygon-line](https://github.com/mapsforge/mapsforge/blob/master/resources/tag-mapping.xsd#L118-L126) in tag-mapping.

**Reply by netman220** [reply]
Hello Emux.
 With osm-tag force-polygon-line="true" everything is fine. Thanks.
Best regards: Alexander Samsiev
